### PR TITLE
Improve pandas and sympy json support

### PR DESCRIPTION
**Pull request by Strilanc**
- Add support for the types needed to make linear combinations in sympy
- Use a stricter pandas equality check in the tests
- Handle pd.Index and pd.MultiIndex recursively
- Fix index type not being preserved and multi-index just not working
- Fix column index name being lost
- Separate testing of external/builtin json round tripping from testing of cirq stuff

Part of #2014

### CODE REVIEW

File: cirq/protocols/json.py [hunk 1]
```diff
@@ -204,14 +212,44 @@ def default(self, o):
...
             return obj_to_dict_helper(o, ['name'], namespace='sympy')
 
+        if isinstance(o, (sympy.Add, sympy.Mul, sympy.Pow)):
+            return obj_to_dict_helper(o, ['args'], namespace='sympy')
+
+        if isinstance(o, sympy.Integer):
+            return {'cirq_type': 'sympy.Integer', 'i': o.p}
+
+        if isinstance(o, sympy.Float):
+            return {'cirq_type': 'sympy.Float', 'approx': float(o)}
```
Comment (Strilanc) [inline cirq/protocols/json.py]: The attributes of sympy float were private, so I didn't dare depend on them.

File: cirq/protocols/json.py [hunk 2]
```diff
@@ -204,14 +212,44 @@ def default(self, o):
...
+            }
+
+        if isinstance(o, pd.MultiIndex):
+            return {
+                'cirq_type': 'pandas.MultiIndex',
+                'tuples': list(o),
+                'names': list(o.names),
+            }
+
+        if isinstance(o, pd.Index):
```
Comment (mpharrigan) [inline cirq/protocols/json.py]: pandas has pretty robust support for json serialization of data-frames. You might want to investigate if you can just use that

Comment (Strilanc) [reply cirq/protocols/json.py]: I looked into it. The `to_json` method on `DataFrame` does not include metadata such as the names of the columns in the index.

```
x = pd.DataFrame(index=pd.MultiIndex.from_tuples([(0, 0)], 
                                                      names=['x', 'y']), 
                      data=[[0, 0]], 
                      columns=pd.Index(['a', 'b'], name='c'))
```

```
print(x)
```

```
c    a  b
x y      
0 0  0  0
```

```
print(x.to_json())
```

```
'{"a":{"[0,0]":0},"b":{"[0,0]":0}}'
```

Comment (mpharrigan) [reply cirq/protocols/json.py]: oh wild

**Pull request merged**